feat: modelling guide (#1221)
ory/docs: docs/keto/modeling/create-permission-model.mdx

---
id: create-permission-model
title: How to create a permission model with Ory Permission Language
sidebar_label: Create a permission model
toc_max_heading_level: 4
---

# Create a permission model

```mdx-code-block
import Mermaid from "@site/src/theme/Mermaid"
```

This section guides you through creating your first permission model using the Ory Permission Language (OPL).

## What is a permission model?

A permission model is a set of rules that define which relations are checked in the database during a permission check.

Permission checks are answered based on:

> - The data available in the database, for example: `User:Bob is owner of Document:X`

Reviewer: `User:Bob is owner of Document:X` - It seemed like you wanted this to be highlighted with Prism but it didn't work. First of all, it works only for code blocks and not for in-line code :( Even when I tried turning this into a code block, the highlighting still didn't work. Probably the way you put the relationship wasn't recognized by prism. Here are examples Patrik added recently: 

````
```keto-relation-tuples
namespace:object#relation:subject
// comment
```

```keto-relationships
namespace:subject is relation of object
is namespace:subject allowed to permission on object?
// comment
```
````

Author: Strange, I don't see see the highlighting either way, not even on https://www.ory.sh/docs/keto/#relationships. @zepatrik is that highlighting broken, or how can I use it?

- Permission rules, for example: "All owners of a document can view it".

> When you ask Ory Permissions: `is User:Bob allowed to view on Document:X`, the system checks up how Bob could have the `view`

Reviewer: Same comment as for line 22

  put into a hierarchy of `folders`. If a user can view the parent `folder`, they can also view all `folders` and `documents` the
  parent `folder` contains.

  ```mdx-code-block
  <Mermaid
    chart={`
  graph LR
      Document -->|parent| Folder
      Folder -->|parent| Folder
      Folder -->|owner| User
      Folder -->|editor| User
      Folder -->|viewer| User
  `}
  />
  ```

This is the complete matrix of relationships presented on a single diagram:

```mdx-code-block
<Mermaid
  chart={`
graph LR
    Document -->|owner| User
    Document -->|editor| User
    Document -->|viewer| User
    Document -->|parent| Folder
    Folder -->|parent| Folder
    Folder -->|owner| User
    Folder -->|editor| User
    Folder -->|viewer| User
`}
/>
```

### Define relation tuples in the OPL

Next, add each relation to the permission config.

:::tip

In Ory Permissions, relationships are declared inside the corresponding class in the Ory Permission Language. Since you store
arbitrary tuples in Ory Permissions, you always have many-to-many relations between these objects. To reflect this in OPL, the

> convention is to pluralize the relation name, for example, `viewers` instead of `viewer`.

Reviewer: I'm not 100% sure I understand `Since you store
arbitrary tuples in Ory Permissions, you always have many-to-many relations between these objects.` 
I think I know what you mean, but... Can you think of a simpler way to say this? 🤔

Author: I can't think of a much simpler way

Author: ```suggestion
In Ory Permissions, relationships are declared inside the corresponding class in the Ory Permission Language. Ory Permissions only
has [many-to-many relationships](<https://en.wikipedia.org/wiki/Many-to-many_(data_model)>) between objects and subjects. To
reflect this in OPL, pluralize the relation name, for example, `viewers` instead of `viewer`.
```

WDYT?

:::

```ts title="permissions-v2.ts"
import { Namespace, Context } from "@ory/keto-namespace-types"

class User implements Namespace {}

class Document implements Namespace {
  // highlight-start
  related: {
    owners: User[]
    editors: User[]
    viewers: User[]
    parents: Folder[]
  }
  // highlight-end
}

class Folder implements Namespace {
  // highlight-start
  related: {
    owners: User[]
    editors: User[]
    viewers: User[]
    parents: Folder[]
  }
  // highlight-end
}
```

### List permissions to check for each object

When you perform an action on behalf of a user, you should check for a specific permission, such as `view` or `edit`, instead of a

> relationship, such as `owner`. Therefore, each permission is derived from relationships and other permissions.

Reviewer: I don't understand where does that `Therefore, each permission is derived from relationships and other permissions.` in the context of this paragraph ;( Is there any way you could rephrase it to make it more obvious?

Author: ```suggestion
relationship, such as `owner`. The concrete permission is then still checked against the relationships. For example, if `owners` can `view` a file, and the `view` permission is checked, then the `owners` relation is looked up in the relationships database.
```

Author: Does that make it clearer?

Reviewer: Yes, I think it's better now.

      //highlight-end
    }
  }

  class Folder implements Namespace {
    related: {
      owners: User[]
      editors: User[]
      viewers: User[]
      parents: Folder[]
    }

    //highlight-start
    permits = {
      delete: (ctx: Context): boolean =>
        this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.delete(ctx)),
      share: (ctx: Context): boolean =>
        this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.share(ctx)),
    }
    //highlight-end
  }
  ```

#### Optional: refactoring

Notice that the permission model you created is hierarchical. This means that everybody who can `view` can also `edit` documents,
and everybody that can `edit` can also `delete` and `share`.

You can refactor the `view` permission like this:

```diff
  view: (ctx: Context): boolean =>
    this.related.viewers.includes(ctx.subject) ||
-   this.related.editors.includes(ctx.subject) ||
-   this.related.owners.includes(ctx.subject) ||
-   this.related.parents.traverse((parent) => parent.permits.view(ctx))
+   this.permits.edit(ctx)
```

> When you apply this to the entire config, you get this code:

Reviewer: It's cool that you showed how to refactor one aspect of the model. How about showing how to refactor something else? Would be cool to have a more detailed look at two things before you just go BAM CONFIG. WDYT?

Author: There is (in the current version of the OPL at least) no more room for refactoring here.

In the future it would be nice to extract `Document` and `Folder` into a common superclass, but that needs more work in the actual Keto code.

Reviewer: got it
